rasterio-stubs/__init__.pyi

```diff
@@ -30,7 +46,7 @@ def open(
     nodata: Nodata | None = ...,
     sharing: bool = ...,
     **kwargs
-) -> Union[DatasetReader, DatasetWriter]: ...
+) -> DatasetReader: ...
```

Comment:
Does this work? Usually you need an [overload for each discriminator](https://docs.python.org/3/library/typing.html#typing.overload), and then a fallback, but maybe it works a little differently in stub files.

```py
@overload
def open(
    fp,
    mode: str = "w",
    ...
    ) -> DatasetWriter: ...

@overload
def open(
    fp,
    mode: str = "r",
    ...
    ) -> DatasetReader: ...

def open(
    fp,
    mode: str = ...,
	...
    ) -> Union[DatasetReader, DatasetWriter]: ...
```

For example, does rasterio support anything other than `w` and `r`? Does it support an append mode?

Reply:
Good question, I was just getting it going for my use case and then figured I'd share the change. I'll take a look and update it to be more comprehensive.